internals/search.py

```diff
@@ -125,40 +126,46 @@ def process_queriable_field(
 # A value that a feature field can be compared against.  It can be
 # a single word or a quoted string.
 VALUE_PATTERN = r'[^" ]+|"[^"]+"'
+# Logical operators.
+# TODO(kyleju): support 'OR' logic
+LOGICAL_OPERATORS_PATTERN = r'AND|OR|NOT'
 
 # Overall, a query term can be either a structured term or a full-text term.
 # Structured terms look like: FIELD OPERATOR VALUE.
 # Full-text terms look like: SINGLE_WORD, or like: "QUOTED STRING".
 TERM_RE = re.compile(
-    '(?P<field>%s)(?P<op>%s)(?P<val>%s)\s+|(?P<textterm>%s)\s+' % (
-        FIELD_NAME_PATTERN, OPERATORS_PATTERN, VALUE_PATTERN,
-        TEXT_PATTERN),
+    '(?P<logical>%s)?\s*(?P<field>%s)(?P<op>%s)(?P<val>%s)\s+|(?P<textterm>%s)\s+' % (
+        LOGICAL_OPERATORS_PATTERN, FIELD_NAME_PATTERN, OPERATORS_PATTERN,
+        VALUE_PATTERN, TEXT_PATTERN),
     re.I)
 
 
 def process_query_term(
-    field_name: str, op_str: str, val_str: str) -> Future:
+    logical_op: str, field_name: str, op_str: str, val_str: str) -> Future:
```

Comment:
I think that only negation can be handled at this level of the code (in `process_query_term()` and `process_queriable_field()`).  Logical OR needs to be handled in `process_query()`.  If there is only the possibility of negation here, it would make more sense to pass it as a boolean keyword like `negate=False`.   But, you might not need it at all, see next comment.

Reply:
I agree and done!